cpp/src/arrow/type.h

```diff
@@ -116,7 +116,7 @@ struct DataType {
 
   bool Equals(const DataType* other) {
     // Call with a pointer so more friendly to subclasses
-    return this == other || (this->type == other->type);
+    return this == other || ((other != nullptr) && (this->type == other->type));
```

Comment:
Fair enough. I am generally comfortable turning a blind eye on the more pedantic points in the interest of productivity